### PR TITLE
Replace `clipboard.js` with native Clipboard API

### LINKED ISSUE #37700: Use native clipboard API instead of a 3rd party dependency
## What problem does this address?
The hook `useCopyToClipboard` uses a 3rd party dependency `clipboard`. We could improve bundle size by using the native clipboard API.

I wrote this PR in an external repo which similarly replaces `clipboard` with the native clipboard API: https://github.com/Automattic/wp-calypso/pull/59266

## What is your proposed solution?

Here's the current code:
https://github.com/WordPress/gutenberg/blob/d754759c45a4f882b01a78e1a2064b77e9d87370/packages/compose/src/hooks/use-copy-to-clipboard/index.js#L44-L68

One could simply use `window.navigator.clipboard.writeText( copyText ).then( ... )` instead of `new Clipboard( ... )`. The only thing I'm unsure about is how to handle the selection clearing and focusing that the existing component uses.

### PR DESCRIPTION
**Pull request by ciampo**
## Description

Closes #37700

Changes in this PR:
- rewrite the `useCopyToClipboard` and the (deprecated) `useCopyOnClick` hooks in `@wordpress/compose` so that they use the native [Clipboard API](https://w3c.github.io/clipboard-apis/#navigator-clipboard)
- remove the `clipboard` dependency from the project's list of dependencies (saving `2.55 kB (23%)` for the `@wordpress/compose`  package)
- add unit tests for both hooks

## How has this been tested?

- Project builds correctly
- Test pass (including the ones added in this PR)
- Usages of the `useCopyToClipboard` hook work as expected (same as production) — see screenshots below for a few examples

## Screenshots 

"Copy" menu item when editing any block in the editor:

https://user-images.githubusercontent.com/1083581/148126016-fa7dfa8b-2bd5-4f39-a14d-909e478fa830.mp4

"Copy URL to clipboard" button when viewing an attachment details in the media library:

image

"Copy URL" button when editing an attachment in the File block:

image

"Copy" button when reviewing a post's info after publishing it:

image

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- N/A I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/compose/src/hooks/use-copy-on-click/index.js [hunk 1]
```diff
@@ -30,45 +25,67 @@ export default function useCopyOnClick( ref, text, timeout = 4000 ) {
...
-			}
+			currentWindow?.navigator?.clipboard
+				?.writeText( textToCopy )
+				.then( () => {
+					// Clear selection and ensure focus stays on the trigger element.
+					if ( trigger ) {
+						currentWindow?.getSelection()?.removeAllRanges();
+
+						trigger.focus();
+					}
```
Comment (ciampo) [inline packages/compose/src/hooks/use-copy-on-click/index.js]: The previous `clipboard.js` implementation used a hack to copy text to the clipboard which involved creating a temporary element, moving focus to it, selecting its text, and executing a "copy" command.

By switching to the new Clipboard APIs, clearing the selection and restoring focus may not be necessary — what do folks think?

File: packages/compose/src/hooks/use-copy-on-click/index.js [hunk 2]
```diff
@@ -30,45 +25,67 @@ export default function useCopyOnClick( ref, text, timeout = 4000 ) {
...
+		if ( typeof ref.current === 'string' ) {
+			// expect `ref` to be a DOM selector
+			triggers = Array.from( document.querySelectorAll( ref.current ) );
+		} else if ( 'length' in ref.current ) {
+			// Expect `ref` to be a `NodeList`
+			triggers = Array.from( ref.current );
+		} else {
+			// Expect `ref` to be a single `Element`
+			triggers = [ ref.current ];
+		}
```
Comment (ciampo) [inline packages/compose/src/hooks/use-copy-on-click/index.js:50]: Previously, the different values for `ref.current` were handles by `Clipboard`'s constructor

File: packages/compose/src/hooks/use-copy-to-clipboard/index.js [hunk 3]
```diff
@@ -39,32 +34,34 @@ export default function useCopyToClipboard( text, onSuccess ) {
...
-			clearSelection();
-			// Handle ClipboardJS focus bug, see
-			// https://github.com/zenorocha/clipboard.js/issues/680
-			node.focus();
+			currentWindow?.navigator?.clipboard
+				?.writeText( textToCopy )
+				.then( () => {
+					// Clear selection and ensure focus stays on the trigger element.
+					currentWindow?.getSelection()?.removeAllRanges();
+					trigger.focus();
```
Comment (ciampo) [inline packages/compose/src/hooks/use-copy-to-clipboard/index.js]: Same question as in [this comment](https://github.com/WordPress/gutenberg/pull/37713/files?diff=split&w=0#r778404668)

**Comment by ciampo** [comment]
Not sure how to fix the Static Analysis CI errors — running `npm install` or `npx lerna bootstrap` doesn't produce any further file changes

**Comment by noahtallen** [comment]
> Not sure how to fix the Static Analysis CI errors — running npm install

Taking a wild guess -- is the npm version the same?

File: packages/compose/src/hooks/use-copy-on-click/index.js [hunk 4]
```diff
@@ -30,45 +25,67 @@ export default function useCopyOnClick( ref, text, timeout = 4000 ) {
...
 	} );
 
-	/** @type {import('react').MutableRefObject<Clipboard | undefined>} */
-	const clipboard = useRef();
 	const [ hasCopied, setHasCopied ] = useState( false );
 
 	useEffect( () => {
 		/** @type {number | undefined} */
 		let timeoutId;
+		/** @type Array<Element>) */
```
Comment (noahtallen) [inline packages/compose/src/hooks/use-copy-on-click/index.js:33]: Out of curiosity, does the `compose` package not support TypeScript directly? I haven't seen this syntax before :) Is it jsdoc?

File: packages/compose/src/hooks/use-copy-on-click/index.js (hunk 1 above)
Comment (noahtallen) [reply packages/compose/src/hooks/use-copy-on-click/index.js]: I agree. If that's the case, we shouldn't need the focus handling. It's probably worth double checking that focus stays on the button after clicking it still :)

File: packages/compose/src/hooks/use-copy-on-click/index.js [hunk 5]
```diff
@@ -1,12 +1,7 @@
-/**
- * External dependencies
- */
-import Clipboard from 'clipboard';
```
Comment (noahtallen) [inline packages/compose/src/hooks/use-copy-on-click/index.js:4]: While the changes in this file are pretty convoluted, it is a deprecated hook after all. The changes in the other hook are a lot more palatable. :P I still think it's worth it to remove the dependency.

File: packages/compose/src/hooks/use-copy-on-click/index.js (hunk 4 above)
Comment (ZebulanStanphill) [reply packages/compose/src/hooks/use-copy-on-click/index.js:33]: https://www.typescriptlang.org/docs/handbook/intro-to-js-ts.html

File: packages/compose/src/hooks/use-copy-on-click/index.js (hunk 5 above)
Comment (ciampo) [reply packages/compose/src/hooks/use-copy-on-click/index.js:4]: I had the same train of thoughts, and came to the same conclusion — refactoring this deprecated hook allows us to remove the dependency from the repo :)

File: packages/compose/src/hooks/use-copy-on-click/index.js (hunk 4 above)
Comment (ciampo) [reply packages/compose/src/hooks/use-copy-on-click/index.js:33]: As Zebulan pointed out, these are type hints for the TypeScript server written with the JSDoc syntax — I decided to adopt this syntax to match other files in the `@wordpress/compose` package — although we should definitely consider migrating the whole package to TypeScript in the future.

**Comment by ciampo** [comment]
> Taking a wild guess -- is the npm version the same?

I noticed that [the `.nvmrc` file only specifies version `14`](https://github.com/WordPress/gutenberg/blob/trunk/.nvmrc), and so when running `nvm use`, since my locally installed  node version was an older `14.x.x`, it didn't update. Then I saw [this commit](https://github.com/WordPress/gutenberg/commit/1714d93487a3301b922f12f425cbc87ce7d7a252), installed specific version `14.18.1` and finally managed to solve the issue.

Anyway, I've also addressed the feedback and added a changelog entry.

**Review by noahtallen — APPROVED** [review]
Tests well for me :+1:

**Comment by jorgefilipecosta** [comment]
Hi @ciampo it seems currently on the trunk the "Copy all content" button does not work at all (nothing happens when clicked, the button is the top options menu). The button to copy a block on the block elipsis menu also is not working. If I revert this PR it starts working again.

**Comment by ciampo** [comment]
> Hi @ciampo it seems currently on the trunk the "Copy all content" button does not work at all (nothing happens when clicked, the button is the top options menu). The button to copy a block on the block elipsis menu also is not working. If I revert this PR it starts working again.

Hey @jorgefilipecosta , thank you for reporting this! Feel free to revert this PR.

Could also open an issue that shows how to reproduce the bug you just described? I'll look into it next week. Thank you!

**Pull request merged**